File: test_runner/test_runner.java
import java.nio.file.*;
import java.util.concurrent.*;
import java.util.stream.*;
import org.json.*;
import java.util.*;
import java.lang.reflect.*;

public class TestRunner {
    private static final Solution solution = new Solution();
    
    private static class TestResult {
        final boolean passed;
        final String error;
        final double timeMs;
        
        TestResult(boolean passed, String error, double timeMs) {
            this.passed = passed;
            this.error = error;
            this.timeMs = timeMs;
        }
    }
    
    private static boolean areEqual(Object actual, Object expected) {
        if (actual == null || expected == null) {
            return actual == expected;
        }
        
        if (actual instanceof int[]) {
            int[] a = (int[]) actual;
            int[] b = (int[]) expected;
            if (a.length != b.length) return false;
            a = Arrays.copyOf(a, a.length);
            b = Arrays.copyOf(b, b.length);
            Arrays.sort(a);
            Arrays.sort(b);
            return Arrays.equals(a, b);
        }
        
        if (actual instanceof List) {
            List<?> actualList = (List<?>) actual;
            List<?> expectedList = (List<?>) expected;
            if (actualList.size() != expectedList.size()) return false;
            
            // Sort if the lists contain comparable elements
            if (!actualList.isEmpty() && actualList.get(0) instanceof Comparable) {
                actualList = new ArrayList<>(actualList);
                expectedList = new ArrayList<>(expectedList);
                Collections.sort((List<Comparable>) actualList);
                Collections.sort((List<Comparable>) expectedList);
            }
            return actualList.equals(expectedList);
        }
        
        return actual.equals(expected);
    }

    private static CompletableFuture<TestResult> runTestAsync(JSONObject testCase, String functionName, int index) {
        return CompletableFuture.supplyAsync(() -> {
            long startTime = System.nanoTime();
            try {
                Method method = null;
                Object[] args = null;
                
                // Get input parameters
                Object input = testCase.get("input");
                if (input instanceof JSONObject) {
                    JSONObject inputObj = (JSONObject) input;
                    args = new Object[inputObj.length()];
                    Class<?>[] paramTypes = new Class<?>[inputObj.length()];
                    
                    int i = 0;
                    for (String key : inputObj.keySet()) {
                        args[i] = inputObj.get(key);
                        paramTypes[i] = args[i].getClass();
                        i++;
                    }
                    method = Solution.class.getMethod(functionName, paramTypes);
                } else {
                    args = new Object[]{input};
                    method = Solution.class.getMethod(functionName, input.getClass());
                }
                
                Object result = method.invoke(solution, args);
                boolean passed = areEqual(result, testCase.get("output"));
                double timeMs = (System.nanoTime() - startTime) / 1_000_000.0;
                
                System.out.printf("Test %2d: %s (%.2fms)%n", 
                    index + 1, 
                    passed ? "\u001B[32m✓\u001B[0m" : "\u001B[31m✗\u001B[0m",
                    timeMs);
                    
                return new TestResult(passed, null, timeMs);
                
            } catch (Exception e) {
                double timeMs = (System.nanoTime() - startTime) / 1_000_000.0;
                String error = e.getCause() != null ? e.getCause().getMessage() : e.getMessage();
                
                System.out.printf("Test %2d: \u001B[31m✗\u001B[0m (Error: %s) (%.2fms)%n", 
                    index + 1, error, timeMs);
                    
                return new TestResult(false, error, timeMs);
            }
        });
    }

    public static void main(String[] args) throws Exception {
        // Read test cases
        String jsonText = new String(Files.readAllBytes(Paths.get("testcases.json")));
        JSONObject data = new JSONObject(jsonText);
        JSONArray cases = data.getJSONArray("test_cases_run");
        String functionName = data.getString("function_name");
        
        System.out.printf("%nRunning tests for function: \u001B[36m%s\u001B[0m%n", functionName);
        System.out.printf("Number of test cases: %d%n%n", cases.length());

        long startTime = System.nanoTime();
        
        List<CompletableFuture<TestResult>> futures = IntStream.range(0, cases.length())
            .mapToObj(i -> runTestAsync(cases.getJSONObject(i), functionName, i))
            .collect(Collectors.toList());

        // Wait for all tests and collect results
        List<TestResult> results = CompletableFuture.allOf(futures.toArray(new CompletableFuture[0]))
            .thenApply(v -> futures.stream()
                .map(CompletableFuture::join)
                .collect(Collectors.toList()))
            .get();
        
        long passed = results.stream().filter(r -> r.passed).count();
        double totalTimeMs = results.stream().mapToDouble(r -> r.timeMs).sum();
        double totalElapsed = (System.nanoTime() - startTime) / 1_000_000.0;
        
        // Print summary
        System.out.println("\nSummary:");
        System.out.println("--------");
        System.out.printf("Passed: %s%d/%d (%.1f%%)%s%n",
            passed == cases.length() ? "\u001B[32m" : "\u001B[31m",
            passed, cases.length(),
            (passed * 100.0 / cases.length()),
            "\u001B[0m");
        System.out.printf("Total time: %.2fms%n", totalElapsed);
        System.out.printf("Average time per test: %.2fms%n", totalTimeMs / cases.length());
        
        System.exit(passed == cases.length() ? 0 : 1);
    }
}

File: test_runner/test_runner.py
import json
import sys
import time
from concurrent.futures import ThreadPoolExecutor, as_completed
from typing import Any, Dict, List, NamedTuple
from colorama import init, Fore, Style
from starter_code import Solution

init()  # Initialize colorama for cross-platform colored output

class TestResult(NamedTuple):
    passed: bool
    error: str
    time_ms: float

def are_equal(actual: Any, expected: Any) -> bool:
    if isinstance(actual, list) and isinstance(expected, list):
        if len(actual) != len(expected):
            return False
        return sorted(actual) == sorted(expected)
    return actual == expected

def run_test_case(solution: Solution, case: Dict, func_name: str, idx: int) -> TestResult:
    start_time = time.time()
    try:
        # Get the function by name and call it
        func = getattr(solution, func_name)
        if not func:
            raise AttributeError(f"Function '{func_name}' not found in Solution class")
            
        # Call the function with the input
        if isinstance(case["input"], dict):
            result = func(**case["input"])
        else:
            result = func(case["input"])
        
        # Compare results
        passed = are_equal(result, case["output"])
        elapsed_ms = (time.time() - start_time) * 1000
        
        # Print result with color
        status = f"{Fore.GREEN}✓{Style.RESET_ALL}" if passed else f"{Fore.RED}✗{Style.RESET_ALL}"
        print(f"Test {idx+1:2d}: {status} ({elapsed_ms:.2f}ms)")
        
        return TestResult(passed, "", elapsed_ms)
        
    except Exception as e:
        elapsed_ms = (time.time() - start_time) * 1000
        print(f"Test {idx+1:2d}: {Fore.RED}✗{Style.RESET_ALL} (Error: {str(e)}) ({elapsed_ms:.2f}ms)")
        return TestResult(False, str(e), elapsed_ms)

def main() -> None:
    try:
        # Load test cases
        with open("testcases.json", "r") as f:
            data = json.load(f)
    except Exception as e:
        print(f"{Fore.RED}Error loading testcases.json: {e}{Style.RESET_ALL}", file=sys.stderr)
        sys.exit(1)

    func_name = data.get("function_name")
    test_cases = data.get("test_cases_run", [])
    
    print(f"\nRunning tests for function: {Fore.CYAN}{func_name}{Style.RESET_ALL}")
    print(f"Number of test cases: {len(test_cases)}\n")

    solution = Solution()
    total_start = time.time()
    
    # Run tests concurrently
    with ThreadPoolExecutor() as executor:
        futures = [
            executor.submit(run_test_case, solution, case, func_name, idx)
            for idx, case in enumerate(test_cases)
        ]
        
        # Collect results
        results = [future.result() for future in as_completed(futures)]
    
    # Calculate statistics
    total_time = (time.time() - total_start) * 1000
    passed = sum(1 for r in results if r.passed)
    avg_time = sum(r.time_ms for r in results) / len(results)
    
    # Print summary with colors
    print(f"\nSummary:")
    print(f"--------")
    print(f"Passed: {Fore.GREEN if passed == len(test_cases) else Fore.RED}"
          f"{passed}/{len(test_cases)} ({passed * 100.0 / len(test_cases):.1f}%)"
          f"{Style.RESET_ALL}")
    print(f"Total time: {total_time:.2f}ms")
    print(f"Average time per test: {avg_time:.2f}ms")

    sys.exit(0 if passed == len(test_cases) else 1)

if __name__ == "__main__":
    main()